src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,380 @@
+                            // (margin was too tight, the shard may end up missing relevant variants)
+                            if (r.getAlignmentEnd()>lastValidPos) {
+                                throw new GATKException("Margin was too tight, a read sticks out by "+(r.getAlignmentEnd()-shard.interval.getEnd())+", going all the way to "+r.getAlignmentEnd());
+                            }
+                            if (r.getAlignmentStart()<firstValidPos) {
+                                throw new GATKException("Margin was too tight, a read starts early by "+(shard.interval.getStart()-r.getAlignmentStart())+", starting at "+r.getAlignmentStart());
+                            }
+                        }
+
+                        while (currentSubShard.getEnd() < r.getStart()) {
+                            if (!readsSoFar.isEmpty()) {
+                                // ship this one.
+                                ContextShard ret = shard.split(currentSubShard).withReads(readsSoFar);
+                                c.output(ret);
+                                readsSoFar = new ArrayList<>();
+                            }
+                            // move to the next shard
+                            currentSubShard = subshards.get(++currentSubShardIndex);
+                        }
+                        // the header slows serialization too much
+                        g.setHeader(null);
+                        readsSoFar.add(g);
+                    }
+                    // done reading, ship what we have
+                    if (!readsSoFar.isEmpty()) {
+                        ContextShard ret = shard.split(currentSubShard).withReads(readsSoFar);
+                        c.output(ret);
+                    }
+                }
+            }
+
+            @Override
+            public void finishBundle(Context c) throws Exception {
+                if (null!=reader) {
+                    reader.close();
+                }
+                super.finishBundle(c);
+            }
+
+            private boolean accept(GATKRead r, SimpleInterval region) {
+                if (r.isUnmapped()) {
+                    return (!r.mateIsUnmapped() && r.getMateStart() >= region.getStart());
+                } else {
+                    // mapped read.
+                    // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                    return r.getStart() >= region.getStart();
+                }
+            }
+
+        };
+    };
+
+    /**
+     * Given a shard that has reads and variants, query Google Genomics' Reference server and get reference info
+     * (including an extra margin on either side), and fill that and the correct variants into readContext.
+     */
+    public static DoFn<ContextShard, ContextShard> fillContext(final ReferenceDataflowSource refSource, final PCollectionView<SAMSequenceDictionary> refDictView, int margin) throws IOException {
```

Comment:
Isn't the `margin` argument ignored now that you're computing the exact reference bases required for the shard?